### PR TITLE
Use ragged dot for full-precision MoE prefill

### PR DESCRIPTION
**Pull request by ry2009**

### CODE REVIEW

**Comment by ry2009** [comment]
code still a bit messy will do pruning tmr

File: lalamo/kernels/ragged_dot.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    group_sizes: Int[Array, " experts"],
+    precision: DotAlgorithmPreset,
+) -> Float[Array, "tokens output_channels"]:
+    return lax.ragged_dot(
+        vectors,
+        expert_weights,
+        group_sizes,
+        precision=precision,
+        out_sharding=sharding_of(vectors),
+    )
```
Comment (knyazer) [inline lalamo/kernels/ragged_dot.py]: just call this one ragged_dot? no need in so many wrappers imo

File: lalamo/model_import/loaders/huggingface.py [hunk 2]
```diff
@@ -480,21 +478,72 @@ def load_moe(
...
+        if (path / "shared_expert" / "up_proj.weight") in weights_dict:
+            if num_shared != 1:
+                raise ValueError("Single shared expert path found but num_shared_experts != 1.")
+            shared_paths = (path / "shared_expert",)
+        elif (path / "shared_experts" / "0" / "up_proj.weight") in weights_dict:
+            shared_paths = tuple(path / "shared_experts" / str(idx) for idx in range(num_shared))
+        else:
+            raise KeyError("Could not find shared expert weights in HF checkpoint.")
+        return stack_experts(module.shared_experts, shared_paths)
+
```
Comment (knyazer) [inline lalamo/model_import/loaders/huggingface.py:540]: check that this works with gpt-oss - just try to uv run lalamo convert {gpt oss} uv run lalamo chat {gpt oss} plz

File: lalamo/modules/linear.py [hunk 3]
```diff
@@ -106,3 +113,25 @@ def __call__(
         if self.biases is not None:
             result = result + self.biases.astype(result.dtype)
         return tuple(jnp.split(result, self.get_split_points(self.output_dims)))
+
+    def call_ragged(
```
Comment (knyazer) [inline lalamo/modules/linear.py:117]: why is this not default impl? shouldn't we always call ragged?

File: lalamo/modules/mlp.py [hunk 4]
```diff
@@ -72,11 +71,18 @@ def _add_moe_expert_outputs(
...
+    index: Int[Array, ""] | Int[Array, " experts"],
+) -> object:
     if not isinstance(leaf, jax.Array):
         return leaf
 
-    out_sharding = sharding_config.make_sharding((None,) * (leaf.ndim - 1))
+    leaf_sharding = sharding_of(leaf)
+    out_sharding = NamedSharding(
+        leaf_sharding.mesh,
+        PartitionSpec(*(None,) * (index.ndim + leaf.ndim - 1)),
```
Comment (knyazer) [inline lalamo/modules/mlp.py]: 🤯 

this took me about 40 seconds to understand, idk if there is a way to rewrite it nicer, but maybe?

File: lalamo/modules/mlp.py [hunk 5]
```diff
@@ -261,26 +271,63 @@ def call_unbatched(
         )
         return result
 
-    def slice_mixture(self, start: int, stop: int) -> Self:
+    def call_ragged_mixture(
```
Comment (knyazer) [inline lalamo/modules/mlp.py:270]: why not default method, why do we need a new method

File: lalamo/modules/mlp.py [hunk 6]
```diff
@@ -331,19 +376,31 @@ def init(self, initializer: Initializer, model_dim: int, hidden_dim: int) -> "Mi
             has_biases=self.router_has_biases,
             is_sharded=False,
         )
-        experts = self.expert_config.init_mixture(
+        routed_experts = self.expert_config.init_mixture(
```
Comment (knyazer) [inline lalamo/modules/mlp.py:375]: tbh i don't like the transition back to routed and shared experts. can we do ragged dot of all the experts, including shared? not only routed

**Comment by knyazer** [comment]
are you done with this @ry2009

File: lalamo/modules/mlp.py [hunk 7]
```diff
@@ -492,61 +528,103 @@ def call_decode_mode(
...
+            forward_pass_config,
+            keychain=keychain,
+        )
+        # ragged_dot leaves the ungrouped padded tail unspecified.
+        weighted_outputs = jnp.where(
+            valid_assignments[:, None],
+            routed_outputs * expert_weights[:, None],
+            0.0,
+        )
+        routed_result = jnp.zeros_like(replicated_inputs).at[token_indices].add(weighted_outputs)
```
Comment (nakhodnov17) [inline lalamo/modules/mlp.py:593]: .at is non-deterministic, and this line causes divergence between reruns under xla_gpu_deterministic_ops=false. Could we replace it with some deterministic implementation?

i.e. (Claude slop):

        inverse_order = jnp.argsort(assignment_order)
        num_tokens, num_channels = replicated_inputs.shape
        per_token_outputs = weighted_outputs.at[inverse_order].get(out_sharding=replicated_matrix_sharding)
        routed_result = per_token_outputs.reshape(num_tokens, num_active, num_channels).sum(axis=1)

**Comment by ry2009** [comment]
Maks is looking at it I'm pretty sure and no not finished will likely work with maks this week or next to finish it @knyazer
